src/time.rs

```diff
@@ -0,0 +1,141 @@
+use std::time::Duration;
+
+pub fn format_duration(duration: &Duration) -> String {
```

Comment:
I guess, you don't actually want to pass `Duration` by a value since it is a `Copy` type.

Reply:
Can we please discuss that more? :) I think I'm missing something

Docs: https://doc.rust-lang.org/std/marker/trait.Copy.html

So as I see it - I can avoid copying if I don't need it because compiler allows me to do it safely, here with `duration: &Duration` I'm passing value by reference which only allocates additional reference that points to same struct

If I change function to `duration: Duration`, whenever I'd call it - a full copy of `Duration` struct will be allocated as well as new reference to it right?

If so, why should I copy if I don't really need a copy here?

Reply:
Well, yeah, you're right. 

Maybe it's just me, my rule of thumb: if I have a `Copy` type then I just pass it by value since it in many cases is more convenient.

> If I change function to duration: Duration, whenever I'd call it - a full copy of Duration struct will be allocated as well as new reference to it right?

Actually it's very compilcated thing : ) But in the end, I think that there will be no difference in the generated machine code, since LLVM very likely will promote this reference to a value and pass it by a copy.

Reply:
(As we say in russian: "Я парень простой, вижу `Copy` тип — передаю его по значению")

Reply:
hahah

Soo, indeed LLVM seems to be able to elide copy, however it's not guaranteed and there are cases where it doesn't do it

If you don't mind I'd keep it as `&Duration` for now, I'll need more Rust experience to get a full sense of when I want implicit copy or not :)

That was super helpful discussion though, really appreciate it :)

Useful links for myself:

- https://stackoverflow.com/questions/38571270/can-rust-optimise-away-the-bit-wise-copy-during-move-of-an-object-someday
- https://www.reddit.com/r/rust/comments/2xxjda/when_should_my_type_be_copy/

Reply:
Heh, I meant the other way around: I bet that LLVM should convert references to a copies, because this usually opens new opportunities for optimizations (IIRC, this is called argument promotion). 

E.g in this code LLVM will promote the reference to the values, two values to be exact: `secs` and `nanos`. And since you don't use nanos in this code it will be elided altogether.

But yeah, LLVM also can elide big copies.

Reply:
wuuut… it indeed can https://github.com/llvm-mirror/llvm/blob/master/lib/Transforms/IPO/ArgumentPromotion.cpp

okay, lemme switch to copy then lol